specs/stein/approved/amd-sev-libvirt-support.rst

```diff
@@ -0,0 +1,367 @@
+each VM with a unique key.  It can also calculate a signature of the
+memory contents, which can be sent to the guest owner as an
+attestation that the memory was encrypted correctly by the firmware.
+Therefore the solution can also provide some degree of protection
+against a malicious hypervisor or compromised cloud operator, although
+it should probably not be relied upon as the sole line of defense.
+
+Use Cases
+---------
+
+#. Booting an instance with a flavor containing an SEV-specific
+   `required trait extra spec
+   <https://docs.openstack.org/nova/latest/user/flavors.html#extra-specs-required-traits>`_
+   will cause that instance to be defined and booted on a compute host
+   which provides that trait.  Additionally the instance should be
+   created with additional XML in the libvirt domain definition which
+   causes QEMU to boot the instance with SEV enabled.
+
+Proposed change
+===============
+
+To achieve the simplest prototype implementation which can possibly
+work, it is proposed that initially during development and testing,
+the placement API would be used to:
+
+- `create a custom trait
+  <https://docs.openstack.org/osc-placement/latest/cli/index.html#trait-create>`_
+  specifically for the purpose of marking flavors as SEV-capable, and
+  then
+
+- `manually assign that trait
+  <https://docs.openstack.org/osc-placement/latest/cli/index.html#resource-provider-trait-set>`_
+  to each compute node which is SEV-capable.
+
+This trait would also indicate that the libvirt driver should include
+extra XML in the guest's domain definition, to enable SEV security,
+ensure that the VM uses machine type ``q35`` and UEFI, and that
+``iommu_platform=on`` is used for all virtio devices, e.g.::
+
+    <domain type='kvm'>
+      <os>
+        <type arch='x86_64' machine='pc-q35-2.11'>hvm</type>
+        <loader readonly='yes' type='pflash'>/usr/share/qemu/ovmf-x86_64-ms-4m-code.bin</loader>
+        <nvram>/var/lib/libvirt/qemu/nvram/sles15-sev-guest_VARS.fd</nvram>
+        <boot dev='hd'/>
+      </os>
+      <launchSecurity type='sev'>
+        <cbitpos>47</cbitpos>
+        <reducedPhysBits>1</reducedPhysBits>
+        <policy>0x0001</policy>
+      </launchSecurity>
+      ...
+    </domain>
+
+For reference, `the AMDSEV GitHub repository
+<https://github.com/AMDESE/AMDSEV/>`_ provides `a complete example of
+SEV-enabled domain XML
+<https://github.com/AMDESE/AMDSEV/blob/master/xmls/sample.xml>`_.
+
+This additional functionality could be implemented within
+``nova.virt.libvirt.LibvirtDriver._set_features()``, e.g. using a new
+``nova.virt.libvirt.LibvirtConfigGuestFeatureSEV`` class.
+
+This work could also mean that images with the property
+``trait:HW_CPU_x86_SEV=required`` would similarly affect the process
+of launching instances.
+
+Immediate follow-up work
+------------------------
+
+It is suggested that follow-up specs be submitted detailing
+the following subsequent work:
```

Comment:
put it in a separate spec then? we need to clarify what is expected to be implemented in "Stein". Is this part of Stein or not? if so then we need to include it here and not as followup work. if we don't then we don't.

Reply:
Actually this is a deliberate manoeuvre on my part - I put it in to preempt any objections which might come on the basis of having merged code which relies on a hardcoded custom trait name.  This is technical debt, and I expect upstream are less likely to worry about that if they can see that the immediate follow-up work to address it has already been carefully thought through and mapped out.  If they say "actually we want all this work to be done in the same spec" then we simply merge this section with the proposed work above.  If they say "that's fine, just move that to a later spec" then again it's easy.  But if I don't include it then it's likely they will say "this introduces technical debt and we don't see any plan to clean up the mess after".

Reply:
Ditto for the technical debt regarding lack of handling of suspend / resume / migrate etc.